File: solutions/ruby/matching-brackets/1/matching_brackets.rb
# Dedicated to Shree DR.MDD for divine inspiration and clarity

class Brackets
  class << self
    def paired? seq
      ledger = []

      seq.chars.each do |sym|
        if initiator? sym
          ledger.push sym
        elsif terminator? sym
          return false unless aligns? ledger.last, sym
          ledger.pop
        end
      end

      ledger.empty?
    end

    private

    def initiator? sym
      /[\[\(\{]/.match? sym
    end

    def terminator? sym
      /[\]\)\}]/.match? sym
    end

    def aligns? lead, trail
      case lead
      when '['
        trail == ']'
      when '('
        trail == ')'
      when '{'
        trail == '}'
      end
    end
  end
end

File: solutions/go/matching-brackets/1/matching_brackets.go
// For my Shree DR.MDD
package brackets

func Bracket(input string) bool {
	match := map[rune]rune{
		']': '[',
		'}': '{',
		')': '(',
	}

	var stk []rune

	for _, ch := range input {
		switch ch {
		case '[', '{', '(':
			stk = append(stk, ch)
		case ']', '}', ')':
			n := len(stk)
			if n == 0 || stk[n-1] != match[ch] {
				return false
			}
			stk = stk[:n-1]
		}
	}
	return len(stk) == 0
}
